src/Libraries/Microsoft.Extensions.Compliance.Abstractions/Redaction/Redactor.cs

```diff
@@ -181,6 +186,11 @@ public int Redact<T>(T value, Span<char> destination, string? format = null, IFo
             return Redact(((IFormattable)value).ToString(format, provider), destination);
         }
 
+        if (value is char[] arrayOfChar)
```

Comment:
I'm not sure whether this is semantically correct behavior. Why should a char array have special treatment over any other kind of array?

I think the test passing an empty array of char may need to be fixed instead.

Reply:
Because the redactor works with sequence of chars or strings represented as `ReadOnlySpan<char>`:
https://github.com/dotnet/extensions/blob/c8e34c57199a2e5e3864649f3bd16258b88ff67e/src/Libraries/Microsoft.Extensions.Compliance.Abstractions/Redaction/Redactor.cs#L28C26-L28C44

Converting an array of char to a string is a wasteful allocation because at the next steps the new string is converted back to a span. 
https://github.com/dotnet/extensions/blob/c8e34c57199a2e5e3864649f3bd16258b88ff67e/src/Libraries/Microsoft.Extensions.Compliance.Abstractions/Redaction/Redactor.cs#L94

Besides, when a user passes `Array.Empty<char>` or `new char[0]`, the user expects the work to be performed on essentially the empty sequence, and not on the string `System.Char[]`.

Reply:
The real issue is the overload resolution. The test is passing a char[] and that's binding to the generic version of the call instead of binding to the overload that takes a ReadOnlySpan. We didn't use to have a generic version of the call, so the test would bind to the ReadOnlySpan overload and things worked.

Reply:
Probably this usage is niche, I am not sure if we need this optimization. I don't think it is wrong though, we already have typechecks for iformattable, and it looks the same IMO.

Reply:
OK, then let's keep it.

@RussKie, please update the code to not use pattern matching and use the same pattern as the if checks for IFormattable and ISpanFormattable above. This pattern is recognized by the JIT and since this is a generic type, the JIT ends up generating code without any of those conditional statements being present. But this only happens when not using pattern matching.

Also, the same change needs to be applied to TryRedact<T>.

Reply:
We can't directly cast, and have to do it via an `object` cast
![image](https://github.com/dotnet/extensions/assets/4403806/a257d36b-f66a-478b-8b63-5323ad4b0385)
![image](https://github.com/dotnet/extensions/assets/4403806/4bc10cab-80cd-493a-abac-9ad9da4646bb)